src/GafferScene/CameraQuery.cpp

```diff
@@ -0,0 +1,551 @@
+	{
+		if( p->parent() == parentPlug )
+		{
+			for( size_t i = 0, eI = parentPlug->children().size(); i < eI; ++i )
+			{
+				if( parentPlug->getChild( i ) == p )
+				{
+					return i;
+				}
+			}
+		}
+		p = p->parent();
+	}
+
+	throw IECore::Exception( "CameraQuery : Plug not in hierarchy." );
+}
+
+} // namespace
+
+GAFFER_NODE_DEFINE_TYPE( CameraQuery )
+
+size_t CameraQuery::g_firstPlugIndex = 0;
+
+CameraQuery::CameraQuery( const std::string &name )
+	:	ComputeNode( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+
+	addChild( new ScenePlug( "scene" ) );
+	addChild( new IntPlug( "cameraMode", Plug::In, (int)CameraMode::Location, (int)CameraMode::Location, (int)CameraMode::RenderCamera ) );
+	addChild( new StringPlug( "location" ) );
+	/// \todo See notes in `ShaderQuery::ShaderQuery`.
+	addChild( new ArrayPlug( "queries", Plug::In, nullptr, 0, std::numeric_limits<size_t>::max(), Plug::Default, false ) );
+	addChild( new ArrayPlug( "out", Plug::Out, nullptr, 0, std::numeric_limits<size_t>::max(), Plug::Default, false ) );
+	addChild( new AtomicCompoundDataPlug( "__internalParameters", Plug::Out, new IECore::CompoundData ) );
+}
+
+CameraQuery::~CameraQuery()
+{
+}
+
+GafferScene::ScenePlug *CameraQuery::scenePlug()
+{
+	return getChild<ScenePlug>( g_firstPlugIndex );
+}
+
+const GafferScene::ScenePlug *CameraQuery::scenePlug() const
+{
+	return getChild<ScenePlug>( g_firstPlugIndex );
+}
+
+Gaffer::IntPlug *CameraQuery::cameraModePlug()
+{
+	return getChild<IntPlug>( g_firstPlugIndex + 1 );
+}
+
+const Gaffer::IntPlug *CameraQuery::cameraModePlug() const
+{
+	return getChild<IntPlug>( g_firstPlugIndex + 1 );
+}
+
+Gaffer::StringPlug *CameraQuery::locationPlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 2 );
+}
+
+const Gaffer::StringPlug *CameraQuery::locationPlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 2 );
+}
+
+Gaffer::ArrayPlug *CameraQuery::queriesPlug()
+{
+	return getChild<ArrayPlug>( g_firstPlugIndex + 3 );
+}
+
+const Gaffer::ArrayPlug *CameraQuery::queriesPlug() const
+{
+	return getChild<ArrayPlug>( g_firstPlugIndex + 3 );
+}
+
+Gaffer::ArrayPlug *CameraQuery::outPlug()
+{
+	return getChild<ArrayPlug>( g_firstPlugIndex + 4 );
+}
+
+const Gaffer::ArrayPlug *CameraQuery::outPlug() const
+{
+	return getChild<ArrayPlug>( g_firstPlugIndex + 4 );
+}
+
+Gaffer::AtomicCompoundDataPlug *CameraQuery::internalParametersPlug()
+{
+	return getChild<AtomicCompoundDataPlug>( g_firstPlugIndex + 5 );
+}
+
+const Gaffer::AtomicCompoundDataPlug *CameraQuery::internalParametersPlug() const
+{
+	return getChild<AtomicCompoundDataPlug>( g_firstPlugIndex + 5 );
+}
+
+Gaffer::StringPlug *CameraQuery::addQuery( const Gaffer::ValuePlug *plug, const std::string &parameter )
+{
+	StringPlugPtr childQueryPlug = new StringPlug( "query0" );
+	childQueryPlug->setValue( parameter );
+
+	ValuePlugPtr newOutPlug = new ValuePlug( "out0", Plug::Out );
+	newOutPlug->addChild( new StringPlug( "source", Plug::Out ) );
```

Comment:
I'm in two minds about this being a StringPlug, since the potential values it can return are constrained like an enum would be. It's more self-explanatory currently to use a string mostly due to deficiencies in using an IntPlug for an enum (where we return an int to Python and not an instance of the enum). Is it worth considering improvements that would allow us to return an enum to Python before we lock in to using a string forever?